common/src/test/kotlin/streams/service/sink/strategy/CUDIngestionStrategyTest.kt

```diff
@@ -0,0 +1,572 @@
+        val nodeEvents = cudQueryStrategy.mergeNodeEvents(list)
+        val nodeDeleteEvents = cudQueryStrategy.deleteNodeEvents(list)
+
+        val relationshipEvents = cudQueryStrategy.mergeRelationshipEvents(list)
+        val relationshipDeleteEvents = cudQueryStrategy.deleteRelationshipEvents(list)
+
+        // then
+        assertEquals(emptyList(), nodeDeleteEvents)
+        assertEquals(emptyList(), relationshipEvents)
+        assertEquals(emptyList(), relationshipDeleteEvents)
+
+        assertEquals(6, nodeEvents.size)
+        assertEquals(10, nodeEvents.map { it.events.size }.sum())
+        val createNodeFooBar = findEventByQuery("""
+                |${StreamsUtils.UNWIND}
+                |CREATE (n:Foo:Bar)
+                |SET n = event.properties
+            """.trimMargin(), nodeEvents)
+        assertEquals(3, createNodeFooBar.events.size)
+        val createNodeFooBarLabel = findEventByQuery("""
+                |${StreamsUtils.UNWIND}
+                |CREATE (n:Foo:Bar:Label)
+                |SET n = event.properties
+            """.trimMargin(), nodeEvents)
+        assertEquals(2, createNodeFooBarLabel.events.size)
+        val mergeNodeFooBar = findEventByQuery("""
+                |${StreamsUtils.UNWIND}
+                |MERGE (n:Foo:Bar {key: event.${CUDIngestionStrategy.ID_KEY}.key})
+                |SET n += event.properties
+            """.trimMargin(), nodeEvents)
+        assertEquals(1, mergeNodeFooBar.events.size)
+        assertNodeEventsContainsKey(mergeNodeFooBar, key)
+        val mergeNodeFooBarLabel = findEventByQuery("""
+                |${StreamsUtils.UNWIND}
+                |MERGE (n:Foo:Bar:Label {key: event.${CUDIngestionStrategy.ID_KEY}.key})
+                |SET n += event.properties
+            """.trimMargin(), nodeEvents)
+        assertEquals(2, mergeNodeFooBarLabel.events.size)
+        assertNodeEventsContainsKey(mergeNodeFooBarLabel, key)
+        val updateNodeFooBar = findEventByQuery("""
+                |${StreamsUtils.UNWIND}
+                |MATCH (n:Foo:Bar {key: event.${CUDIngestionStrategy.ID_KEY}.key})
+                |SET n += event.properties
+            """.trimMargin(), nodeEvents)
+        assertEquals(1, updateNodeFooBar.events.size)
+        assertNodeEventsContainsKey(updateNodeFooBar, key)
+        val updateNodeFooBarLabel = findEventByQuery("""
+                |${StreamsUtils.UNWIND}
+                |MATCH (n:Foo:Bar:Label {key: event.${CUDIngestionStrategy.ID_KEY}.key})
+                |SET n += event.properties
+            """.trimMargin(), nodeEvents)
+        assertEquals(1, updateNodeFooBarLabel.events.size)
+        assertNodeEventsContainsKey(updateNodeFooBarLabel, key)
+    }
+
+    @Test
+    fun `should delete nodes`() {
+        // given
+        val detachMarkers = listOf(1, 3, 8, 10)
+        val list = (1..10).map {
+            val labels = if (it % 2 == 0) listOf("Foo", "Bar") else listOf("Foo", "Bar", "Label")
```

Comment:
In your test cases, please make sure to use labels with garbage data and also properties with the same, because this will happen in the real world.   For example:

labels = ["WellBehaved", "C̸r̵a̵z̵y̵ ̶.̵ ̶ ̴ ̸ ̶ ̶ ̵ ̴L̴a̵b̸e̶l̴"]
properties = ["simple", "not..... SO SIMPLE!"]

Reply:
Ok